### PR TITLE
build pr commit on pr build

### PR DESCRIPTION
**Pull request by garyverhaegen-da**
# Open for debate

> This PR is the answer to "how would we do it", please read further for _what_ "it" is, why we'd want it, and why we wouldn't want it.

I am obviously biased towards making this change (otherwise I would not have made this PR), so there may be more arguments against than I am representing here.

## What does this do?

By default, when Azure Pipelines builds a pull request, what it runs is the result of merging the PR branch into master. By taking the second parent of the current commit (`HEAD^2`), this resets the current working directory to be just the unadulterated pull request as it can be seen on GitHub.

## Why would we *not* want this?

Though I would deplore the fact that this Azure Pipelines behaviour is undocumented and does not expose any kind of control over it, I have to agree that, in many cases, running CI builds on the result of merging the PR into master _is_ the right thing to do. Running the PR branch alone (as this change would do) is not quite enough to be confident about the final status of merging it into master.

## Why, then, _would_ we want this?

The problem is, this breaks down somewhat in our case, because master just moves too fast. Or because Azure does not go far enough along its own line of reasoning, depending on your point of view.

In a world where master moves sufficiently slowly, it is a good thing that looking at a PR's build result gives you confidence merging it will result in a working master _because the PR build result is already testing the result of the merge_. However, in a world, like ours, where, for the vast majority of PRs, between the time the Azure build starts and the time the PR would get merged, _master has changed_, this behaviour results in two undesirable consequences:
1. If we know about this Azure behaviour, the PR build result is misleading because master has changed since the build has run; therefore, a successful build gives us a false sense of security.
2. If we do not know about this Azure behaviour, the build result is misleading because it simply has not run what we believe it has run.

In my opinion, the default behaviour of Azure would really only make sense if it followed its own logic all the way through and restarted all the PR builds whenever master changes. Note that we already have a way to approximate that behaviour (with much smarter resource utilization) in the `automerge-after-rebase` tag.

### CODE REVIEW

**Comment by cocreature** [comment]
tl;dr: I prefer the current behavior.

You are completely right that in a world where master moves fast, testing the PR merged into master does not ensure that your PR will still build once it is merged. I also agree than in an ideal world, CI would be sufficiently fast that rebasing every PR (or rerunning the PR merged into the latest master) whenever master changes to make sure that PRs actually build when merged would be great. However, we tried that for a while and in my opinion it didn’t work for us and I don’t think anything has changed that would make it a viable option now.

So that leaves us with two options:
1. (Proposed behavior in this PR) Test the PR on the branch without forcing a rebase or merge. This can end up testing a state that was arbitrarily old either because someone had a long-running branch that they didn’t rebase or because someone maliciously based their branch on an old state (not too worried about that personally but something to keep in mind).
2. (Current behavior) Test the PR merged into master at the time the PR is opened/has last changed.

I have seen a few PRs that had legitimate failures since master had changed and were caught by the fact that we use 2. While it might be confusing, I would argue that causing a bit of confusion for the PR author (which is usually a one-time thing until you are aware of the behavior) is far better than merging a broken PR. I have only seen one PR that passed CI _because_ we use 2 but shouldn’t have passed: a release PR that checks the list of changed files. It looks like we could use the same trick you use here to get the list of changed files for the actual PR to prevent this from happening for release PRs which as far as I know are the only ones that care about the list of changed files.
